[cxx-interop] Add SIL function representation cxx_method; Support extending C++ types.

zoecarver opened the pull request:
There are three major changes here:
 1. The addition of `SILFunctionTypeRepresentation::CXXMethod`.
 2. C++ methods are imported with their members *last*. Then the arguments are switched when emitting the IR for an application of the function.
 3. Clang decls are now marked as foreign witnesses.

These are all steps towards being able to have C++ protocol conformance.

A few other notes:
* There is a surprising lack of C++ method tests. As a separate PR, I'll add some shortly. 
* This will also let me write some tests for the witness layout optimizations (as [requested here](https://github.com/apple/swift/pull/32291#discussion_r532613678)).
* Protocol conformance is also needed for some future generics work.
* I'm hoping to submit future patches to add support for static C++ methods and operators. 

Resolves [SR-12750](https://bugs.swift.org/browse/SR-12750).

varungandhi-apple commented on include/swift/AST/ExtInfo.h:
```diff
@@ -167,6 +167,9 @@ enum class SILFunctionTypeRepresentation : uint8_t {
   /// A closure invocation function that has not been bound to a context.
   Closure,
+
+  /// A C++ method that takes a "self" argument (not a static C++ method or cosntructor).
```
Typo: "cosntructor" -> constructor.

varungandhi-apple commented on lib/AST/ASTMangler.cpp:
```diff
@@ -1693,6 +1693,7 @@ void ASTMangler::appendImplFunctionType(SILFunctionType *fn) {
       OpArgs.push_back('B');
       appendClangTypeToVec(OpArgs);
       break;
+    case SILFunctionTypeRepresentation::CXXMethod:
```
Please update `docs/ABI/Mangling.rst` for this addition.

varungandhi-apple commented on lib/SIL/IR/SILFunctionType.cpp:
```diff
@@ -4233,6 +4245,7 @@ TypeConverter::getLoweredFormalTypes(SILDeclRef constant,
     bridgedResultType = resultType;
     break;
 
+  case SILFunctionTypeRepresentation::CXXMethod:
```
In the body of the case, you could simplify the code to check if the representation is `CXXMethod` instead of doing `auto method = dyn_cast_or_null<clang::CXXMethodDecl>(constant.getDecl()->getClangDecl())`.

zoecarver replied:
We need `method`, so we'd end up casting it anyway, so I'm not sure how much better it would be. But happy to update if you'd still like me to.

varungandhi-apple commented on lib/SIL/IR/Bridging.cpp:
```diff
@@ -187,6 +188,7 @@ Type TypeConverter::getLoweredCBridgedType(AbstractionPattern pattern,
     case SILFunctionType::Representation::Thin:
     case SILFunctionType::Representation::Method:
     case SILFunctionType::Representation::ObjCMethod:
+    case SILFunctionTypeRepresentation::CXXMethod:
```
Not part of this PR, but the comment above looks a tad outdated.

```
// Functions that are already represented as blocks or C function pointers
// don't need bridging.
```

Potentially want to update that (or remove it) since you are updating nearby stuff.

zoecarver replied:
Hmm. Is this comment saying "don't need to be bridged here" or "don't need bridging in general"? The latter is outdated but the former seems OK. Also happy to remove it.

varungandhi-apple commented on lib/IRGen/GenCall.cpp:
```diff
@@ -3246,14 +3268,18 @@ static void externalizeArguments(IRGenFunction &IGF, const Callee &callee,
+  } else if (callee.getRepresentation() ==
+             SILFunctionTypeRepresentation::CXXMethod) {
+    // Skip the "self" param.
+    paramEnd--;
+    params = params.drop_front();
```
This reads in a very confusing manner because dropping from the front makes it look like the start should be updated and not the end. Is this logically equivalent to the `firstParam += 1` handling in the block case? That doesn't modify `params`, so I'm not sure why we are modifying `params` here.

zoecarver replied:
Heh, this is why we need some more C++ method tests. The `params.drop_front()` part isn't supposed to be there. Remember, the self parameter is now last. We could drop the last param, but that doesn't seem necessary because we're going to have to update `paramEnd` anyway. 

Maybe `paramCount` would be a better name, though.

zoecarver commented:
@varungandhi-apple all resolved. Thanks for the review!

varungandhi-apple commented on docs/ABI/Mangling.rst:
```diff
@@ -531,8 +531,8 @@ Types
   FUNCTION-KIND ::= 'L'                      // objc block function type with canonical C type (escaping) (DWARF only; otherwise use 'B' or 'zB' C-TYPE)
-  FUNCTION-KIND ::= 'C'                      // C function pointer type
-  FUNCTION-KIND ::= 'zC' C-TYPE              // C function pointer type with with non-canonical C type
+  FUNCTION-KIND ::= 'C'                      // C function pointer / method type
+  FUNCTION-KIND ::= 'zC' C-TYPE              // C function pointer / method type with with non-canonical C type
```
Please use "C++ method" (otherwise, it's not clear if this is referring to ObjC methods or C++ methods).

zoecarver replied:
Thanks, done.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please test Windows.

varungandhi-apple commented:
Mostly LGTM, deferring to other folks for specifics.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@rjmccall and @hlopko friendly ping.

hyp commented on lib/IRGen/GenCall.cpp:
```diff
@@ -2187,7 +2191,12 @@ class SyncCallEmission final : public CallEmission {
     case SILFunctionTypeRepresentation::Block:
-      adjusted.add(getCallee().getBlockObject());
+    case SILFunctionTypeRepresentation::CXXMethod:
+      if (getCallee().getRepresentation() ==
```
Why are you handling the block and cxxmethod in same case if you need to do a different call for block vs cxx method? I think it would be better to have separate cases.

zoecarver replied:
Aha. Turns out there was actually a reason for this: we want the fall-through here for both cases. Without this weird if-inside-the-case pattern we don't `externalizeArguments` for `::Block`.

hyp commented on lib/SILGen/SILGenType.cpp:
```diff
@@ -417,9 +417,18 @@ template<typename T> class SILGenWitnessTable : public SILWitnessVisitor<T> {
-            requirementRef, requirementRef.withDecl(witness.getDecl()),
-            witness);
+        auto newDecl = requirementRef.withDecl(witness.getDecl());
+        // Don't import Objective-C methods as forien. If the following
```
Typo: foreign.

hyp commented on lib/SILGen/SILGenType.cpp:
```diff
@@ -417,9 +417,18 @@ template<typename T> class SILGenWitnessTable : public SILWitnessVisitor<T> {
+        // It will be imported as the following type:
+        //   () -> NSString
+        // But the first is correct, so make sure we don't mark this witness
+        // as forien.
```
Typo: foreign.

hyp commented on lib/SILGen/SILGenType.cpp:
```diff
@@ -417,9 +417,18 @@ template<typename T> class SILGenWitnessTable : public SILWitnessVisitor<T> {
-            witness);
+        auto newDecl = requirementRef.withDecl(witness.getDecl());
+        // Don't import Objective-C methods as forien. If the following
+        // Objective-C function is imported here as foreing:
```
Do you mean a C++ method here instead of Objective-C function?

hyp commented on lib/Serialization/Serialization.cpp:
```diff
@@ -3906,6 +3906,7 @@ static uint8_t getRawStableSILFunctionTypeRepresentation(
   SIMPLE_CASE(SILFunctionTypeRepresentation, ObjCMethod)
+  SIMPLE_CASE(SILFunctionTypeRepresentation, CXXMethod)
   SIMPLE_CASE(SILFunctionTypeRepresentation, WitnessMethod)
   SIMPLE_CASE(SILFunctionTypeRepresentation, Closure)
```
The new case should go after `Closure` to match the enum above, right?

zoecarver committed:
[cxx-interop] Add SIL function representation cxx_method; Support extending C++ types.

There are three major changes here:
    1. The addition of "SILFunctionTypeRepresentation::CXXMethod".
    2. C++ methods are imported with their members *last*. Then the arguments are switched when emitting the IR for an application of the function.
    3. Clang decls are now marked as foreign witnesses.

These are all steps towards being able to have C++ protocol conformance.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

zoecarver commented:
@swift-ci please smoke test.

slavapestov commented on lib/SILGen/SILGenType.cpp:
```diff
@@ -415,8 +415,19 @@ template<typename T> class SILGenWitnessTable : public SILWitnessVisitor<T> {
+        // as foreign.
+        if (dyn_cast_or_null<clang::CXXMethodDecl>(
+                witness.getDecl()->getClangDecl()))
+          newDecl = newDecl.asForeign();
```
Why are C++ methods different than Objective-C methods in this regard?

zoecarver replied:
Basically, if an Objective-C method is marked as foreign it will be imported using the Objective-C type and not the Swift type. For example, it might be imported as `() -> NSString` instead of `() -> String` (which is what we want).

hyp reviewed (approved):
LGTM
